### ISSUE TITLE
Get rid of <?>

### ISSUE THREAD

**Issue by matozoid** [javaparser/javaparser#758]
Some node types are generic, only to indicate the return type of the setters. This is useless for the user who has to add `<?>` to all these types. `Type` avoids this with a trick: the setter for annotations returns `Type` and the inherting classes like `UnionType` override it casting the return type to the correct one.
```java
@Override
public UnionType setAnnotations(NodeList<AnnotationExpr> annotations) {
    return (UnionType) super.setAnnotations(annotations);
}
```

This concerns `TypeDeclaration`, `BodyDeclaration`, and `ReferenceType`.

Although this breaks the API, I don't think anybody will complain about having to remove this crap.

**Comment by jeandersonbc** [comment]
I agree that there isn't much benefit of Generics in this circumstance. Not to mention that some classes using `ReferenceType` don't really need to know about the concrete type of `ReferenceType`. Adding `<?>` is just annoying for these cases.

The suggested approach is cleaner and also will reduce the raw type warnings from the compiler.

I would like to help with this one.
Could you assign it for me?

**Comment by matozoid** [comment]
Consider yourself assigned.

**Comment by matozoid** [comment]
`CallableDeclaration` has been added to the list :( :(

**Comment by arturbosch** [comment]
But actually MethodDeclaration and ConstructorDeclaration should be excluded from the list :). Were the getNodeList methods generated?

**Comment by matozoid** [comment]
They weren't on the list :( And yes, those are generated, and I plan on deprecating them in favour of using the metamodel.

**Comment by matozoid** [comment]
#770 for @arturbosch !

**Comment by matozoid** [comment]
Okay, there is a disadvantage here: the type parameter is there only for the return type of the setters. Now, if I want to remove it, I have to move all the implemented `NodeWith...` interfaces down from the abstract nodes to the concrete nodes, because there I can definitely tell the interface what setter return type they should have.

But that means you can't use those interface on the abstract node types anymore. I wonder if this is worth it.

@ftomassetti ?

**Comment by matozoid** [comment]
... I may be able to fix that if I get rid of the type parameter on the `NodeWith...` interface too, and simply depend on what's-it-called, return type reduction or so.

**Comment by ftomassetti** [comment]
Could you give an example, just to be sure I do not misunderstand? I think that the higher in the hierarchy we use the NodeWith interfaces the better it is. I agree that the type returned by the setters is not _that_ important

**Comment by matozoid** [comment]
Yeah, I guessed I was being unclear :)

Current:
```java
interface NodeWithSomething<T> {
  T setSomething(s);
}

class TypeDeclaration<T> implements NodeWithSomething<T> {
  private Something something;

  // This T is the only reason we have all those type parameters:
  T setSomething(s){...} 
}
class COID extends TypeDeclaration<COID> { 
}
```

My first attempt to remove the `<?>` from `TypeDeclaration`:

```java
interface NodeWithSomething<T> {
  T setSomething(s);
}

class TypeDeclaration {
  private Something something;

  // Leave the implementation here, but with a wide return type
  TypeDeclaration setSomething(s){...} 
}

class COID extends TypeDeclaration implements NodeWithSomething<COID> { 
  // Add the interface in the concrete class and narrow the return type.
  COID setSomething(s){return(COID)super.setSomething} 
}
```
As you can see, I need to move the interface.
If I also implement it on TypeDeclaration as NodeWithSomething<TypeDeclaration>,
the compiler complains that I shouldn't reimplement it with a different type parameter in COID.

Idea:
```java
// Take away the generics completely.
interface NodeWithSomething {
  Node setSomething(s);
}

class TypeDeclaration implements NodeWithSomething {
  private Something something;

  // Narrow the type a little
  TypeDeclaration setSomething(s){...} 
}

class COID extends TypeDeclaration {
  // Now narrow it completely
  COID setSomething(s){return(COID)super.setSomething} 
}
```

I think this last one should work...

**Comment by ftomassetti** [comment]
And anyway you have to redeclare the actual method in the concrete class to implement it, right? So I guess it makes sense

**Comment by matozoid** [comment]
Yeah, the type parameter made the return type narrowing automatic. I feel a little bad about all the super-calls, but at least we don't bother people with odd `<?>`'s anymore.

**Comment by ftomassetti** [comment]
I think we should focus on the user point of view. We can privately suffer for the super-calls for the sake of the greater benefit

**Comment by matozoid** [comment]
Uhuh, but I'm also thinking of stacktraces with with several identical set methods above each other :-O
